tests/unit/s2n_self_talk_alerts_test.c

```diff
@@ -151,27 +160,29 @@ int main(int argc, char **argv)
 
     /* Test that we ignore Warning Alerts in S2N_ALERT_IGNORE_WARNINGS mode in TLS1.2 */
     {
-        EXPECT_NOT_NULL(config = s2n_config_new());
-        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
-        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
-
         /* Create a pipe */
         struct s2n_test_io_pair io_pair;
         EXPECT_SUCCESS(s2n_io_pair_init(&io_pair));
 
-        /* Set up the callback to send an alert after receiving ClientHello */
-        struct alert_ctx warning_alert = { .write_fd = io_pair.server, .invoked = 0, .count = 2, .level = TLS_ALERT_LEVEL_WARNING, .code = TLS_ALERT_UNRECOGNIZED_NAME };
-        EXPECT_SUCCESS(s2n_config_set_client_hello_cb(config, client_hello_send_alerts, &warning_alert));
-
         /* Create a child process */
         pid = fork();
         if (pid == 0) {
             /* This is the client process, close the server end of the pipe */
             EXPECT_SUCCESS(s2n_io_pair_close_one_end(&io_pair, S2N_SERVER));
 
+            EXPECT_OK(cleanup(&cert_chain_pem, &private_key_pem, &chain_and_key));
```

Comment:
"The cleanup() function (DEFER_CLEANUP) doesnt get called if the process exists abruptly"
Couldn't you change the process so that it doesn't exit abruptly? Like with `return` instead of `exit()`?

Reply:
The `fork` creates a separate process where we run client specific code. The original process run server specific code.

The work "abruptly" doesnt imply unintentional; `mock_client` is all of the client specific code and once we are done we want to call `exit` so that we kill the process. If we call `return` then we will continue to call server specific code and that would be UB. Here is a diagram:

```
main {
  common_code

  fork {
    client_specific_code
    exit
  }

  server_specific_code
}
```

The solution I have opted for is to avoid any allocation in the `common_code` section and therefore avoid cleanup. However, when we have multiple tests with common init then it can get verbose/require large refactor. I tried to avoid a large refactor here but can make a common init method if that would be preferred.

Reply:
I was thinking something like this would make the DEFER_CLEANUP work:
```
mock_client() {
    return 0;
}

main {
    pid = fork();
    if (pid == 0) {
        mock_client();
        return 0;
    }
}

// server code
```

Reply:
the reason we need `exit` is because we use that to propagate the status of the client process; by `return`ing we essentially mask any potential errors in the client process. In server_specific_code we have the following check:

```
EXPECT_EQUAL(waitpid(-1, &status, 0), pid);
EXPECT_EQUAL(status, 0);
```